File: backend/services/plot_service.py
from sqlalchemy.orm import Session
from sqlalchemy import text, func, and_
from models import LandPlot, PlotOrder
from typing import Optional, Dict, Any, List
import json
import logging

logger = logging.getLogger(__name__)

class PlotService:
    
    def get_all_plots_geojson(self, db: Session) -> Dict[str, Any]:
        """Get all plots as GeoJSON FeatureCollection"""
        try:
            logger.info("Fetching all plots as GeoJSON")
            # Query plots with geometry as GeoJSON
            query = text("""
                SELECT 
                    id::text,
                    plot_code,
                    status,
                    area_hectares,
                    district,
                    ward,
                    village,
                    attributes,
                    created_at,
                    updated_at,
                    ST_AsGeoJSON(geometry)::json as geometry
                FROM land_plots
                ORDER BY plot_code
            """)
            
            result = db.execute(query)
            plots = result.fetchall()
            
            logger.info(f"Found {len(plots)} plots in database")
            
            features = []
            for plot in plots:
                # Validate geometry
                if not plot.geometry:
                    logger.warning(f"Plot {plot.plot_code} has no geometry, skipping")
                    continue
                    
                feature = {
                    "type": "Feature",
                    "properties": {
                        "id": plot.id,
                        "plot_code": plot.plot_code,
                        "status": plot.status,
                        "area_hectares": float(plot.area_hectares),
                        "district": plot.district,
                        "ward": plot.ward,
                        "village": plot.village,
                        "attributes": plot.attributes or {},
                        "created_at": plot.created_at.isoformat() + "Z",
                        "updated_at": plot.updated_at.isoformat() + "Z"
                    },
                    "geometry": plot.geometry
                }
                features.append(feature)
            
            logger.info(f"Returning {len(features)} valid plot features")
            return {
                "type": "FeatureCollection",
                "features": features
            }
            
        except Exception as e:
            logger.error(f"Error fetching plots as GeoJSON: {e}")
            raise
    
    def get_plot_geojson(self, db: Session, plot_id: str) -> Optional[Dict[str, Any]]:
        """Get single plot as GeoJSON Feature"""
        try:
            query = text("""
                SELECT 
                    id::text,
                    plot_code,
                    status,
                    area_hectares,
                    district,
                    ward,
                    village,
                    attributes,
                    created_at,
                    updated_at,
                    ST_AsGeoJSON(geometry)::json as geometry
                FROM land_plots
                WHERE id = :plot_id
            """)
            
            result = db.execute(query, {"plot_id": plot_id})
            plot = result.fetchone()
            
            if not plot:
                return None
            
            return {
                "type": "Feature",
                "properties": {
                    "id": plot.id,
                    "plot_code": plot.plot_code,
                    "status": plot.status,
                    "area_hectares": float(plot.area_hectares),
                    "district": plot.district,
                    "ward": plot.ward,
                    "village": plot.village,
                    "attributes": plot.attributes or {},
                    "created_at": plot.created_at.isoformat() + "Z",
                    "updated_at": plot.updated_at.isoformat() + "Z"
                },
                "geometry": plot.geometry
            }
            
        except Exception as e:
            logger.error(f"Error fetching plot {plot_id} as GeoJSON: {e}")
            raise
    
    def get_plot_by_id(self, db: Session, plot_id: str) -> Optional[LandPlot]:
        """Get plot by ID"""
        try:
            return db.query(LandPlot).filter(LandPlot.id == plot_id).first()
        except Exception as e:
            logger.error(f"Error fetching plot {plot_id}: {e}")
            raise
    
    def search_plots(
        self,
        db: Session,
        district: Optional[str] = None,
        ward: Optional[str] = None,
        village: Optional[str] = None,
        status: Optional[str] = None,
        min_area: Optional[float] = None,
        max_area: Optional[float] = None,
        bbox: Optional[str] = None
    ) -> Dict[str, Any]:
        """Search plots with various filters"""
        try:
            # Build WHERE conditions
            conditions = []
            params = {}
            
            if district:
                conditions.append("LOWER(district) = LOWER(:district)")
                params["district"] = district
            
            if ward:
                conditions.append("LOWER(ward) = LOWER(:ward)")
                params["ward"] = ward
            
            if village:
                conditions.append("LOWER(village) = LOWER(:village)")
                params["village"] = village
            
            if status:
                conditions.append("status = :status")
                params["status"] = status
            
            if min_area is not None:
                conditions.append("area_hectares >= :min_area")
                params["min_area"] = min_area
            
            if max_area is not None:
                conditions.append("area_hectares <= :max_area")
                params["max_area"] = max_area
            
            if bbox:
                try:
                    minx, miny, maxx, maxy = map(float, bbox.split(','))
                    conditions.append("""
                        ST_Intersects(
                            geometry, 
                            ST_MakeEnvelope(:minx, :miny, :maxx, :maxy, 4326)
                        )
                    """)
                    params.update({
                        "minx": minx, "miny": miny, 
                        "maxx": maxx, "maxy": maxy
                    })
                except ValueError:
                    logger.warning(f"Invalid bbox format: {bbox}")
            
            where_clause = "WHERE " + " AND ".join(conditions) if conditions else ""
            
            query = text(f"""
                SELECT 
                    id::text,
                    plot_code,
                    status,
                    area_hectares,
                    district,
                    ward,
                    village,
                    attributes,
                    created_at,
                    updated_at,
                    ST_AsGeoJSON(geometry)::json as geometry
                FROM land_plots
                {where_clause}
                ORDER BY plot_code
            """)
            
            result = db.execute(query, params)
            plots = result.fetchall()
            
            features = []
            for plot in plots:
                feature = {
                    "type": "Feature",
                    "properties": {
                        "id": plot.id,
                        "plot_code": plot.plot_code,
                        "status": plot.status,
                        "area_hectares": float(plot.area_hectares),
                        "district": plot.district,
                        "ward": plot.ward,
                        "village": plot.village,
                        "attributes": plot.attributes or {},
                        "created_at": plot.created_at.isoformat() + "Z",
                        "updated_at": plot.updated_at.isoformat() + "Z"
                    },
                    "geometry": plot.geometry
                }
                features.append(feature)
            
            return {
                "type": "FeatureCollection",
                "features": features
            }
            
        except Exception as e:
            logger.error(f"Error searching plots: {e}")
            raise
    
    def get_system_stats(self, db: Session) -> Dict[str, Any]:
        """Get system statistics"""
        try:
            # Plot statistics
            plot_stats = db.execute(text("""
                SELECT 
                    COUNT(*) as total_plots,
                    COUNT(*) FILTER (WHERE status = 'available') as available_plots,
                    COUNT(*) FILTER (WHERE status = 'taken') as taken_plots,
                    COUNT(*) FILTER (WHERE status = 'pending') as pending_plots,
                    COUNT(DISTINCT district) as districts,
                    COUNT(DISTINCT ward) as wards,
                    COUNT(DISTINCT village) as villages,
                    SUM(area_hectares) as total_area_hectares
                FROM land_plots
            """)).fetchone()
            
            # Order statistics
            order_stats = db.execute(text("""
                SELECT 
                    COUNT(*) as total_orders,
                    COUNT(*) FILTER (WHERE status = 'pending') as pending_orders,
                    COUNT(*) FILTER (WHERE status = 'approved') as approved_orders,
                    COUNT(*) FILTER (WHERE status = 'rejected') as rejected_orders
                FROM plot_orders
            """)).fetchone()
            
            return {
                "total_plots": plot_stats.total_plots,
                "available_plots": plot_stats.available_plots,
                "taken_plots": plot_stats.taken_plots,
                "pending_plots": plot_stats.pending_plots,
                "total_orders": order_stats.total_orders,
                "pending_orders": order_stats.pending_orders,
                "approved_orders": order_stats.approved_orders,
                "rejected_orders": order_stats.rejected_orders,
                "districts": plot_stats.districts,
                "wards": plot_stats.wards,
                "villages": plot_stats.villages,
                "total_area_hectares": float(plot_stats.total_area_hectares or 0)
            }
            
        except Exception as e:
            logger.error(f"Error fetching system stats: {e}")
            raise

File: src/services/plotService.ts
import { Plot, OrderData, Order } from '../types/land';

const API_BASE = import.meta.env.VITE_API_URL;
fetch(`${API_BASE}/plots`);

// helper to normalize status from API → union type
function normalizeStatus(status: string | null | undefined): "available" | "taken" | "pending" {
  if (status === "available" || status === "taken" || status === "pending") {
    return status;
  }
  return "pending"; // fallback if API sends unexpected value
}

class PlotService {
  private async fetchWithErrorHandling(url: string, options?: RequestInit): Promise<Response> {
    try {
      console.log(`Making API request to: ${url}`);
      
      const response = await fetch(url, {
        ...options,
        headers: {
          'Content-Type': 'application/json',
          ...options?.headers,
        },
      });

      if (!response.ok) {
        const errorText = await response.text();
        console.error(`API Error ${response.status}:`, errorText);
        
        let errorMessage = `HTTP ${response.status}`;
        try {
          const errorJson = JSON.parse(errorText);
          errorMessage = errorJson.detail || errorJson.message || errorMessage;
        } catch {
          errorMessage = errorText || errorMessage;
        }
        
        throw new Error(errorMessage);
      }

      return response;
    } catch (error) {
      console.error('Network error:', error);
      
      if (error instanceof TypeError && error.message.includes('fetch')) {
        throw new Error('Unable to connect to the server. Please check your internet connection and try again.');
      }
      
      throw error;
    }
  }

  async getAllPlots(): Promise<Plot[]> {
    try {
      console.log('Fetching all plots from API...');
      const response = await this.fetchWithErrorHandling(`${API_BASE}/api/plots`);
      const data = await response.json();
      
      console.log('Raw API response:', data);
      
      if (!data || typeof data !== 'object') {
        throw new Error('Invalid response format from server. Please check if the backend is running.');
      }
      
      if (data.type !== 'FeatureCollection') {
        throw new Error('Expected GeoJSON FeatureCollection format. The API may be returning incorrect data.');
      }
      
      if (!Array.isArray(data.features)) {
        throw new Error('Invalid features array in response. The shapefile data may not be properly imported.');
      }
      
      if (data.features.length === 0) {
        throw new Error('No land plots found. Please run the database seed script to import shapefile data.');
      }
      
      // Transform GeoJSON features to Plot objects
      const plots: Plot[] = data.features.map((feature: any, index: number) => {
        try {
          if (!feature.properties || !feature.geometry) {
            console.warn(`Feature ${index} missing properties or geometry:`, feature);
            return null;
          }
          
          const props = feature.properties;
          
          // Validate required properties
          if (!props.id || !props.plot_code) {
            console.warn(`Feature ${index} missing required properties:`, props);
            return null;
          }
          
          // Validate geometry
          if (!feature.geometry.coordinates || feature.geometry.coordinates.length === 0) {
            console.warn(`Feature ${index} has invalid geometry:`, feature.geometry);
            return null;
          }
          
          return {
            id: props.id,
            plot_code: props.plot_code,
            status: normalizeStatus(props.status),
            area_hectares: parseFloat(props.area_hectares) || 0,
            district: props.district || 'Unknown',
            ward: props.ward || 'Unknown',
            village: props.village || 'Unknown',
            geometry: feature.geometry,
            attributes: props.attributes || {},
            created_at: props.created_at || new Date().toISOString(),
            updated_at: props.updated_at || new Date().toISOString(),
          };
        } catch (error) {
          console.error(`Error processing feature ${index}:`, error, feature);
          return null;
        }
      }).filter(Boolean) as Plot[]; // Remove null entries safely
      
      console.log(`Successfully processed ${plots.length} plots from ${data.features.length} features`);
      
      if (plots.length === 0) {
        console.warn('No valid plots found in API response');
      }
      
      return plots;
      
    } catch (error) {
      console.error('Error fetching plots:', error);
      throw new Error(`Failed to load land plots: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  async getPlotById(plotId: string): Promise<Plot | null> {
    try {
      console.log(`Fetching plot ${plotId}...`);
      const response = await this.fetchWithErrorHandling(`${API_BASE}/api/plots/${plotId}`);
      const feature = await response.json();
      
      if (!feature.properties || !feature.geometry) {
        return null;
      }
      
      const props = feature.properties;
      return {
        id: props.id,
        plot_code: props.plot_code,
        status: normalizeStatus(props.status),
        area_hectares: parseFloat(props.area_hectares),
        district: props.district,
        ward: props.ward,
        village: props.village,
        geometry: feature.geometry,
        attributes: props.attributes || {},
        created_at: props.created_at,
        updated_at: props.updated_at,
      };
      
    } catch (error) {
      console.error(`Error fetching plot ${plotId}:`, error);
      return null;
    }
  }

  async createOrder(plotId: string, orderData: OrderData): Promise<Order> {
    try {
      console.log(`Creating order for plot ${plotId}:`, orderData);
      
      // Validate order data
      if (!orderData.first_name?.trim()) {
        throw new Error('First name is required');
      }
      
      if (!orderData.last_name?.trim()) {
        throw new Error('Last name is required');
      }
      
      if (!orderData.customer_phone?.trim()) {
        throw new Error('Customer phone is required');
      }
      
      if (!orderData.customer_email?.trim()) {
        throw new Error('Customer email is required');
      }
      
      const response = await this.fetchWithErrorHandling(
        `${API_BASE}/api/plots/${plotId}/order`,
        {
          method: 'POST',
          body: JSON.stringify(orderData),
        }
      );
      
      const order = await response.json();
      console.log('Order created successfully:', order);
      
      return order;
      
    } catch (error) {
      console.error(`Error creating order for plot ${plotId}:`, error);
      throw new Error(`Failed to create order: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  async searchPlots(filters: {
    district?: string;
    ward?: string;
    village?: string;
    status?: string;
    min_area?: number;
    max_area?: number;
    bbox?: string;
  }): Promise<Plot[]> {
    try {
      const params = new URLSearchParams();
      
      Object.entries(filters).forEach(([key, value]) => {
        if (value !== undefined && value !== null && value !== '') {
          params.append(key, value.toString());
        }
      });
      
      const url = `${API_BASE}/api/plots/search?${params.toString()}`;
      console.log(`Searching plots with filters:`, filters);
      
      const response = await this.fetchWithErrorHandling(url);
      const data = await response.json();
      
      if (data.type !== 'FeatureCollection') {
        throw new Error('Expected GeoJSON FeatureCollection format');
      }
      
      const plots: Plot[] = data.features.map((feature: any) => {
        const props = feature.properties;
        return {
          id: props.id,
          plot_code: props.plot_code,
          status: normalizeStatus(props.status),
          area_hectares: parseFloat(props.area_hectares),
          district: props.district,
          ward: props.ward,
          village: props.village,
          geometry: feature.geometry,
          attributes: props.attributes || {},
          created_at: props.created_at,
          updated_at: props.updated_at,
        };
      });
      
      console.log(`Found ${plots.length} plots matching search criteria`);
      return plots;
      
    } catch (error) {
      console.error('Error searching plots:', error);
      throw new Error(`Failed to search plots: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  async getSystemStats(): Promise<{
    total_plots: number;
    available_plots: number;
    taken_plots: number;
    pending_plots: number;
    total_orders: number;
    districts: number;
    wards: number;
    villages: number;
    total_area_hectares: number;
  }> {
    try {
      console.log('Fetching system statistics...');
      const response = await this.fetchWithErrorHandling(`${API_BASE}/api/stats`);
      const stats = await response.json();
      
      console.log('System stats:', stats);
      return stats;
      
    } catch (error) {
      console.error('Error fetching system stats:', error);
      throw new Error(`Failed to fetch statistics: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  async getOrders(filters?: {
    status?: string;
    plot_id?: string;
    limit?: number;
    offset?: number;
  }): Promise<{
    orders: Order[];
    total: number;
  }> {
    try {
      const params = new URLSearchParams();
      
      if (filters) {
        Object.entries(filters).forEach(([key, value]) => {
          if (value !== undefined && value !== null && value !== '') {
            params.append(key, value.toString());
          }
        });
      }
      
      const url = `${API_BASE}/api/orders?${params.toString()}`;
      console.log('Fetching orders...');
      
      const response = await this.fetchWithErrorHandling(url);
      const data = await response.json();
      
      console.log(`Fetched ${data.orders?.length || 0} orders`);
      return data;
      
    } catch (error) {
      console.error('Error fetching orders:', error);
      throw new Error(`Failed to fetch orders: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  // Health check method
  async checkHealth(): Promise<boolean> {
    try {
      const response = await fetch(`${API_BASE}/health`, {
   
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      });
      
      return response.ok;
    } catch (error) {
      console.error('Health check failed:', error);
      return false;
    }
  }
}

// Export singleton instance
export const plotService = new PlotService();
export default plotService;